fluent/src/langneg.js

```diff
@@ -0,0 +1,246 @@
+ *
+ *    What it does ([requested] * [available] = [supported]):
+ *
+ *    ['en-US'] * ['en'] = ['en']
+ *
+ *  2) likely subtags from LDML 4.3 Likely Subtags has been added
+ *
+ *    The most obvious likely subtag that can be computed is a duplication
+ *    of the language field onto region field (`fr` => `fr-FR`).
+ *
+ *    On top of that, likely subtags may use a list of mappings, that
+ *    allow the algorithm to handle non-obvious matches.
+ *    For example, making sure that we match `en` to `en-US` or `sr` to
+ *    `sr-Cyrl`, while `sr-RU` to `sr-Latn-RU`.
+ *
+ *    This list can be taken directly from CLDR Supplemental Data.
+ *
+ *    What it does ([requested] * [available] = [supported]):
+ *
+ *    ['fr'] * ['fr-FR'] = ['fr-FR']
+ *    ['en'] * ['en-US'] = ['en-US']
+ *    ['sr'] * ['sr-Latn', 'sr-Cyrl'] = ['sr-Cyrl']
+ *
+ *  3) region range check has been added
+ *
+ *    Lastly, the last form of check is against the requested locale ID
+ *    but with the region field replaced with a `*` range.
+ *
+ *    The rationale here laid out in LDML 4.4 Language Matching:
+ *      "(...) normally the fall-off between the user's languages is
+ *      substantially greated than regional variants."
+ *
+ *    In other words, if we can't match for the given region, maybe
+ *    we can match for the same language/script but other region, and
+ *    it will in most cases be preferred over falling back on the next
+ *    language.
+ *
+ *    What it does ([requested] * [available] = [supported]):
+ *
+ *    ['en-AU'] * ['en-US'] = ['en-US']
+ *    ['sr-RU'] * ['sr-Latn-RO'] = ['sr-Latn-RO'] // sr-RU -> sr-Latn-RU
+ */
+function filterMatches(requestedLocales, availableLocales) {
+  const supportedLocales = [];
+
+  outer:
+  for (const requestedLocale of requestedLocales) {
+
+    for (const availableLocale of availableLocales) {
+      if (requestedLocale === availableLocale) {
+        supportedLocales.push(availableLocale);
+        continue outer;
+      }
+    }
+
+    for (const availableLocale of availableLocales) {
+      if (compareLocales(requestedLocale, availableLocale)) {
+        supportedLocales.push(availableLocale);
+        continue outer;
+      }
+    }
+
+    const maxRequestedLocale = addLikelySubtags(requestedLocale);
+
+    if (maxRequestedLocale) {
+      for (const availableLocale of availableLocales) {
+        if (compareLocales(maxRequestedLocale, availableLocale)) {
+          supportedLocales.push(availableLocale);
+          continue outer;
+        }
+      }
+    }
+
+    const regionRange = getRegionRange(maxRequestedLocale || requestedLocale);
+
+    for (const availableLocale of availableLocales) {
+      if (compareLocales(regionRange, availableLocale)) {
+        supportedLocales.push(availableLocale);
+        continue outer;
+      }
+    }
+  }
+
+  return supportedLocales;
+}
+
+export function negotiateLanguages(requestedLocales,
```

Comment:
Thanks for bringing this up. For now, each package in the `fluent.js` repository has its docstrings extracted using the `documentation.js` package. The extracted markdown files (one per package) land in the top-loevel `docs/` directory.

It might be better to move the docs into each package's directory as `$(PACKAGE)/docs/api.md`, in case we want to add more hand-written docs later on.